src/light_data/light_store.rs

```diff
@@ -0,0 +1,404 @@
+// This is a temporary shim which should be merged with scalar_store
+// Currently it only exists for reading store-dumps
+
+use crate::field::FWrap;
+#[cfg(not(target_arch = "wasm32"))]
+use proptest::prelude::*;
+#[cfg(not(target_arch = "wasm32"))]
+use proptest_derive::Arbitrary;
+
+use anyhow::anyhow;
+use std::collections::BTreeMap;
+
+use crate::light_data::Encodable;
+use crate::light_data::LightData;
+use crate::scalar_store::ScalarExpression;
+use crate::scalar_store::ScalarStore;
+use crate::store::ScalarPointer;
+use crate::store::ScalarPtr;
+use crate::sym::Sym;
+use crate::tag::ExprTag;
+
+use crate::field::LurkField;
+
+#[derive(Debug, PartialEq)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Arbitrary))]
+#[cfg_attr(not(target_arch = "wasm32"), proptest(no_bound))]
+/// LightStore contains a fragment of the ScalarStore, but using the `LightExpr` type
+pub struct LightStore<F: LurkField> {
+    /// An analogous to the ScalarStore's scalar_map, but with `LightExpr` instead of `ScalarExpression`
+    pub scalar_map: BTreeMap<ScalarPtr<F>, Option<LightExpr<F>>>,
+}
+
+impl<F: LurkField> Encodable for LightStore<F> {
+    fn ser(&self) -> LightData {
+        // TODO: this clone is loathsome
+        self.scalar_map
+            .clone()
+            .into_iter()
+            .collect::<Vec<(ScalarPtr<F>, Option<LightExpr<F>>)>>()
+            .ser()
+    }
+    fn de(ld: &LightData) -> anyhow::Result<Self> {
+        let pairs = Vec::<(ScalarPtr<F>, Option<LightExpr<F>>)>::de(ld)?;
+        Ok(LightStore {
+            scalar_map: pairs.into_iter().collect(),
+        })
+    }
+}
+
+impl<F: LurkField> LightStore<F> {
+    fn insert_scalar_string(
+        &self,
+        ptr0: ScalarPtr<F>,
+        store: &mut ScalarStore<F>,
+    ) -> anyhow::Result<String> {
+        let mut s = String::new();
+        let mut tail_ptrs = vec![];
+        let mut ptr = ptr0;
+        let strnil_ptr = ScalarPtr::from_parts(ExprTag::Str.as_field(), F::zero());
+
+        // TODO: this needs to bail on encountering an opaque pointer
+        while let Some(LightExpr::StrCons(c, cs)) = self.get(&ptr).flatten() {
+            let chr = c
+                .value()
+                .to_char()
+                .ok_or_else(|| anyhow!("Non-char head in LightExpr::StrCons"))?;
+            store.insert_scalar_expression(c, Some(ScalarExpression::Char(chr)));
+            s.push(chr);
+            if cs != strnil_ptr {
+                tail_ptrs.push(cs);
+            }
+            ptr = cs
+        }
+        // Useful when called from insert_scalar_symbol
+        if s.is_empty() {
+            return Err(anyhow!("encountered no StrCons in LightStore::insert_scalar_string, is this a tring pointer?"));
+        }
+
+        // If we've already inserted this string, no need to go through suffixes again
+        if let Some(ScalarExpression::Str(old_value)) = store
```

Comment:
Not for now, but a question: if the string is already inserted, can/should we assume all the individual chars have been too (if needed)? That might be a valid assumption if either it must be true algorithmically, or if we don't need it to be in order to be satisfied with the result? (The same question goes for symbols and constituent strings below.)

Reply:
`insert_scalar_string` maintains the invariant that if it is called on a `&ptr` that resolves to `Some(LightExpression::StrCons(foo, bar))` then it will insert all suffixes of the string starting at `ptr`, and all its constituent characters.
`insert_scalar_symbol` maintains the invariant that if it is called on a `&ptr` that resolves to `Some(LightExpression::SymCons(foo, bar))` then it will insert all suffixes of the path starting at `ptr`, and call `insert_scalar_string` on all its consitutent symbol fragment names.

Reply:
Since chars are immediate pointers we can actually never need to insert them at all! Like, since you can always build a `ScalarExpr::Char` from its pointer (because the hash of char is itself), they don't need their own entries

Reply:
@johnchandlerburnham in the future world where `LightStore` and `ScalarStore` have fused and we model immediate pointers in the LightExpr AST, certainly. 

But not for the specific scope of this PR, which we have circumscribed (on Zulip) to:
- the `LightStore` does not contain `Char` pointers, reflecting the "new" world,
- the `ScalarStore` needs to contain them, reflecting the "old world".